### ISSUE TITLE
interface to Collier

### ISSUE THREAD

**Issue by wkotlarski** [FlexibleSUSY/FlexibleSUSY#64]
I'd like to kick off the discussion about the interface to `Collier`. `Collier` doesn't solve the problem raised in #25 but is definitely an improvement over `LoopTools`. As you might know there is no `C/C++` interface to `Collier`, we will have to write it ourself. Here's one solution I came up with.

On the Fortran side, one has an intermediate module where we use the Fortran2003 feature of `iso_c_binding` to wrap the functions from `Collier`
```fortran
subroutine A0_cll_I(m0) bind(c)
   use COLLIER
   use iso_c_binding
   double complex A0_cll_I
   double complex, intent(in) :: m0
   call A0_cll(A0_cll_I,m0)
end subroutine f
```
On the `C++` side we then have to declare those functions as `extern` to avoid name mangling. I wasn't able the get the stuff compiling without this intermediate `Fortran` module. As far as I understand, the `Fortran` compiler does append certain things to the function name. It's not as bad as what the `C++` compiler does, but it still changes the function name. The `bind(C)` guarantees that the name will be the name of the function in lowercase (you can also give the name you want as an extra argument to `bind`). `iso_c_bind` also provides mapping of `C` and `Fortran` data types. 

It should be possible to call `Collier` functions without this intermediate step, but the procedure might be compiler dependent. On the other hand, using my approach we'd have to wrap every function from `Collier`. That's something like 20 functions, so it's not tragic. Thoughts?

**Comment by Expander** [comment]
I'd recommend to write a FORTRAN wrapper function for each loop function from COLLIER. As you say, this enables a high degree of compatibility with multiple FORTRAN/C++ compilers.

For the interface itself, I'd recommend to

* use a FORTRAN `function`, instead of a `subroutine`, because a `subroutine` returns `void`
* use `bind(C,name='A0_cll_I')` to ensure that the compiler generates the right function name `A0_cll_I`

At the C++ level we then also need wrapper routines with a sane interface, for example:
```
extern "C" {
   double A0_cll_I(const double*); // FORTRAN wrapper routine
}

namespace collier {
   /// C++ wrapper with non-pointer parameter
   double A0(double x)
   {
      const double m = x;
      return A0_cll_I(&m);
   }
}
```

**Comment by wkotlarski** [comment]
@Expander, can you prepare the configure script to link against `Collier`? How should we handle the selection of loop library?

**Comment by Expander** [comment]
Yes, I can do that. However, I'd like to take the opportunity to change the way the loop function library is chosen, because currently we already have a few options, and it we would now introduce another one:

Currently we have the options `--enable-fflite` and `--enable-looptools`, where only one of them must be selected (i.e. they are conflicting). Adding another (conflicting) flag `--enable-collier` would probably not be very elegant.

So, I'd propose to replace the options `--enable-fflite` and `--enable-looptools` by:
```
./configure --with-loopfunction-library=<lflib>
```
where `<lflib>` is either `softsusy`, `fflite`, `looptools` or `collier`.
What do you think?

For backwards compatibility we could still keep the `--enable-fflite` and `--enable-looptools` flags, if needed.

**Comment by wkotlarski** [comment]
And what do you say about giving the possibility to select loop libraries at run time?

**Comment by wkotlarski** [comment]
@Expander, @azedarach added C functions to `pv`. So `FlexibleDecay` is calling `LoopTools` through this wrapper. Are you ok with this? If yes I'll send you a pull request. I'm afraid that this change might interfere with the `collier` interface we talked about.

**Comment by wkotlarski** [comment]
I'd like to continue the discussion on the implementation of other loop libraries. I'm in favor of being able to switch loop libraries at run time. What comes to my mind is something along the factory design pattern. Anyone has anything against it? @Expander might object because of virtual dispatch.

**Comment by wkotlarski** [comment]
Here's a prefilimenary idea for an abstraction layer for one-loop libraries. For simplicity, imagine that there's only one loop function, called `B0` that doesn't take any arguments. Then the loop interface could look like
```cpp
enum class LoopLibrary {
   LoopTools,
   Collier
};

class LoopLibraryInterface {
   public:
      static LoopLibraryInterface *get_library(LoopLibrary);
      virtual double B0() = 0;
};


class LoopTools : public LoopLibraryInterface {
   public:
      double B0() {
         cout << "LoopTools\n";
         return 1.; 
      }   
};

class Collier : public LoopLibraryInterface {
   public:
      double B0() {
         cout << "Collier\n";
         return 1.; 
      }   
};

LoopLibraryInterface* LoopLibraryInterface::get_library(LoopLibrary type) {
   switch (type) {
      case LoopLibrary::Collier:
         return new Collier;
      case LoopLibrary::LoopTools:
         return new LoopTools;
   }   
}
```
The function `get_library` returns a pointer to a one loop library, e.g.
```cpp
LoopLibraryInterface* lib = LoopLibraryInterface::get_library(LoopLibrary::Collier);
```
After that, the `B0` function can be called as `lib->B0()`. The `FlexibleSUSY` will not care about which subclass of `LoopLibraryInterface` is `lib` a member of.

**Comment by Expander** [comment]
Thanks for the draft!

Some technical points:

* I don't like the idea of passing ownership of the library type to the user via `get_library()`. One better option would be to return a `std::uniqe_ptr`.
* Returning a pointer would be meaningful only if you want to store the library for later usage and/or you want to express that the result could be `0`. Do we want this behaviour? If not, an alternative would be [Scott Meyer's singleton](https://stackoverflow.com/questions/1008019/c-singleton-design-pattern/1008289#1008289) implementation, which returns a reference (which can never be `0` and is more tricky to be stored in a class).
* What would be the behaviour if one of the libraries has not been configured in? Should `get_library(type)` throw an exception if `type` is not configured in?
* It may be useful to provide a default type that always works / is always available. This may be FFLITE or the SOFTSUSY loop functions.

**Comment by wkotlarski** [comment]
This is clearly not my are of expertise. The first bullet point could be addressed like so
```cpp
class LoopLibraryInterface {
   public:
      static std::unique_ptr<LoopLibraryInterface> get_library(LoopLibrary);
      virtual double B0() = 0;
};

std::unique_ptr<LoopLibraryInterface> LoopLibraryInterface::get_library(LoopLibrary type) {
   switch (type) {
      case LoopLibrary::Collier:
         return std::make_unique<Collier>();
      case LoopLibrary::LoopTools:
         return std::make_unique<LoopTools>();
   }   
}
```
If i understand correctly, the advantage is that we will not be able to make a copy of this pointer. But i guess you'd rather have this singleton so that there's only one object at all.

Points 3 & 4 are up for discussion. We could handle it at configure level and then just secure it by throwing an exception from `get_library`.

**Comment by wkotlarski** [comment]
@uukhas wants to have an access to Collier so this issue becomes urgent. I've implemented a working call to C00 function in a3c8fe7976b2727e55d7190252535c01a8ee91d6. The interface idea is as we discussed above There are two intermediate levels: the fortran 2003 module that prepares Collier functions to be called from c++ and an actual c++ `Collier` class that inherits from `Loop_library_interface`. Please check if you agree with an overall setup and the naming of files/directories/classes. The naming of files, classes and subdirectories is maybe the most important as it's the biggest pain to change once we move much further. If we agree on this one example, wrapping other functions is purely repetitive work. I've want to keep the naming and arguments for loop functions as in `pv.cpp`. For now this code should coexist with `pv.cpp`. @Expander, we'll need your help in modifying a configure script. For now I just modified `Makefile` by hand. 

We need to:
- compile file `collier_wrapper.f03` as library, linking it with `-lcollier`. There's also a possibility that user might have compiled Collier as a static library. Should we support both options?
- compile classes derived from `Loop_library_interface`
- add our `collier_wrapper` lib to FS linker

Hope it's all fine for you all.

**Comment by wkotlarski** [comment]
For completeness:

The fortran interface can be compiles as
```
gfortran -std=f2008 -c collier_wrapper.f03 -I path/to/collier/include/dir
ar crv libcollier_wrapper.a collier_wrapper.o
```
And add to LOOPFUNCLIBS in `Makefile`
```
src/loop_libraries/libcollier_wrapper.a -L/path/to/Collier/lib/folder -lcollier
```
